data_transfer/schemas/record.py

```diff
@@ -57,3 +57,28 @@ def download_folder(self) -> Path:
         )
         path.mkdir(parents=True, exist_ok=True)
         return path
+
+    # ensure validators are run when changing values, not just construction
+    class Config:
+        validate_assignment = True
+
+    @validator("is_processed")
+    def after_downloaded(cls, v: bool, values: dict) -> bool:
+        # values: a dict containing the name-to-value mapping of any previously-validated fields
+        # NOTE: only checks if the value 'v' is set to True
+        if v and not values["is_downloaded"]:
+            raise ValueError("NOT ALLOWED: this Record is not downloaded yet.")
+        return v
+
+    @validator("is_prepared", always=True)
+    def after_processed(cls, v: bool, values: dict) -> bool:
+        if v and not values["is_processed"]:
+            raise ValueError("NOT ALLOWED: this Record is not processed yet.")
+        return v
+
+    @validator("is_uploaded")
+    def after_prepared(cls, v: bool, values: dict) -> bool:
+        # NOTE: this value is normally set after upload; somewhat 'too little to late'
+        if v and not values["is_prepared"]:
+            raise ValueError("NOT ALLOWED: this Record is not prepared yet.")
+        return v
```

Comment:
Given the logic is the same across these three validators, I would pull it out into a public static method of the Record class, which should be accessible via `cls`.

```python
   @staticmethod
    def value_exists_or_raise(value_exists: bool, record: dict, value: str) -> bool:
        if value_exists and not record[value]:
            _type = value.split("-")[0]
            raise ValueError(f"NOT ALLOWED: this Record is not {_type} yet.")
        return value_exists
```

Then each validator method can call it (I assume) via:

```python
return cls.value_exists_or_raise(v, values, "is_downloaded")
```

This also means you can test the if statement -- so two extra tests: one for fail and one for success.